lib/charms/operator_libs_linux/v0/dnf.py

```diff
@@ -0,0 +1,462 @@
+        """
+        return self.__data["name"]
+
+    @property
+    def arch(self) -> Optional[str]:
+        """Get architecture of package.
+
+        Returns:
+            (Optional[str]):
+                Architecture of package. Defaults to None if package is absent.
+        """
+        return self.__data.get("arch", None)
+
+    @property
+    def epoch(self) -> Optional[str]:
+        """Get epoch of package.
+
+        Returns:
+            (Optional[str]):
+                Epoch of package. Defaults to None if package is absent.
+        """
+        return self.__data.get("epoch", None)
+
+    @property
+    def version(self) -> Optional[str]:
+        """Get version of package.
+
+        Returns:
+            (Optional[str]):
+                Version number of package. Defaults to None if package is absent.
+        """
+        return self.__data.get("version", None)
+
+    @property
+    def fullversion(self) -> Optional[str]:
+        """Get full version of package.
+
+        Returns:
+            (Optional[str]):
+                Full version of package. Defaults to None if package is absent.
+        """
+        if self.absent:
+            return None
+
+        return (
+            f"{(f'{self.epoch}:' if self.epoch else '')}"
+            f"{self.version}{(f'-{self.release}' if self.release else '')}"
+        )
+
+    @property
+    def release(self) -> Optional[str]:
+        """Get release of package.
+
+        Returns:
+            (Optional[str]):
+                Release of package. Defaults to None if package is absent.
+        """
+        return self.__data.get("release", None)
+
+    @property
+    def repo(self) -> Optional[str]:
+        """Get repository package is from.
+
+        Returns:
+            (Optional[str]):
+                Repository of package: Defaults to None if package is absent.
+        """
+        return self.__data.get("repo", None)
+
+
+def _load_cache(
+    cache_type: Literal["installed", "available"]
+) -> List[Dict[str, Union[str, _PackageState]]]:
+    """Load DNF cache information.
+
+    Args:
+        cache_type (Literal["installed", "available"]):
+            Cache information category to load.
+
+    Raises:
+        DNFExecutionError: Raised if incorrect cache type is passed.
+
+    Returns:
+        (List[Dict[str, Union[str, _PackageState]]]):
+            Loaded cache information.
+    """
+    if cache_type not in ["installed", "available"]:
+        raise DNFExecutionError(
+            f"Cache type must be either installed or available, not {cache_type}"
+        )
+
+    result = []
+    state = {"installed": _PackageState.INSTALLED, "available": _PackageState.AVAILABLE}
+    for pkg_name, pkg_version, pkg_repo in itertools.filterfalse(
+        lambda x: len(x) != 3,
+        [
+            line.strip("\n").strip().split()
+            for line in StringIO(
+                _dnf("list", postargs=[f"--{cache_type}"]).stdout.strip("\n"),
```

Comment:
`dnf list ...` output is more than likely intended to be human-readable, however, it does not have messages like `apt` about being an unstable interface. I am using its output because it is easy to parse with regexs, custom parsers, etc. dnf does have a `packages.db` file, but I have not figured out what the schema is.

dnf itself is written with Python (I do think dnf v5 is being written in C++ however), so it has utilities for interacting with `packages.db`, but the dnf Python package is only available on RPM-based distributions. I am avoidant towards using that library because this charm library will more than likely live inside charms that can also be deployed to Ubuntu bases.